ecdsa/src/lib.rs

```diff
@@ -214,39 +229,92 @@ where
 where
+    C: Curve + CheckSignatureBytes,
     SignatureSize<C>: ArrayLength<u8>,
 {
     type Error = Error;
 
     fn try_from(bytes: &[u8]) -> Result<Self, Error> {
-        if bytes.len() == <SignatureSize<C>>::to_usize() {
-            Ok(Self {
-                bytes: GenericArray::clone_from_slice(bytes),
-            })
-        } else {
-            Err(Error::new())
+        if bytes.len() != <SignatureSize<C>>::to_usize() {
+            return Err(Error::new());
         }
+
+        let bytes = GenericArray::clone_from_slice(bytes);
+        C::check_signature_bytes(&bytes)?;
+
+        Ok(Self { bytes })
     }
 }
 
-impl<C> From<asn1::Signature<C>> for Signature<C>
+impl<C> TryFrom<asn1::Signature<C>> for Signature<C>
 where
-    C: Curve,
+    C: Curve + CheckSignatureBytes,
     C::FieldSize: Add + ArrayLength<u8>,
     asn1::MaxSize<C>: ArrayLength<u8>,
     <C::FieldSize as Add>::Output: Add<asn1::MaxOverhead> + ArrayLength<u8>,
 {
-    fn from(doc: asn1::Signature<C>) -> Signature<C> {
-        let mut bytes = SignatureBytes::<C>::default();
+    type Error = Error;
+
+    fn try_from(doc: asn1::Signature<C>) -> Result<Signature<C>, Error> {
+        let mut bytes = GenericArray::default();
         let scalar_size = C::FieldSize::to_usize();
         let r_begin = scalar_size.checked_sub(doc.r().len()).unwrap();
         let s_begin = bytes.len().checked_sub(doc.s().len()).unwrap();
 
         bytes[r_begin..scalar_size].copy_from_slice(doc.r());
         bytes[s_begin..].copy_from_slice(doc.s());
-        Signature { bytes }
+
+        C::check_signature_bytes(&bytes)?;
+        Ok(Signature { bytes })
+    }
+}
+
+/// Ensure a signature is well-formed.
+pub trait CheckSignatureBytes: Curve
+where
+    SignatureSize<Self>: ArrayLength<u8>,
+{
+    /// Validate that the given signature is well-formed.
+    ///
+    /// This trait is auto-impl'd for curves which impl the
+    /// `elliptic_curve::Arithmetic` trait, which validates that the
+    /// `r` and `s` components of the signature are in range of the
+    /// scalar field.
+    ///
+    /// Note that this trait is not for verifying a signature, but allows for
+    /// asserting properties of it which allow infallible conversions
+    /// (e.g. accessors for the `r` and `s` components)
+    fn check_signature_bytes(bytes: &SignatureBytes<Self>) -> Result<(), Error> {
+        // Ensure `r` and `s` are both non-zero
+        for scalar_bytes in bytes.chunks(Self::FieldSize::to_usize()) {
+            if scalar_bytes.iter().all(|&b| b != 0) {
+                return Err(Error::new());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<C> CheckSignatureBytes for C
+where
+    C: Curve + Arithmetic,
+    SignatureSize<C>: ArrayLength<u8>,
+{
+    /// When curve arithmetic is available, check that the scalar components
+    /// of the signature are in range.
+    fn check_signature_bytes(bytes: &SignatureBytes<C>) -> Result<(), Error> {
```

Comment:
The reason why this checks the signature as a whole rather than the individual components is it might make sense in some cases to have special-case handling of the individual components. The main case that comes to mind is EIP-2098 signatures. But this may be of dubious value, and it may still make more sense to have a "check scalar in range" trait which can be available even if a full curve arithmetic implementation is not.